cmd/syncthing/generate/generate.go

```diff
@@ -0,0 +1,147 @@
+}
+
+func (c *CLI) Run() error {
+	log.SetFlags(0)
+
+	if c.HideConsole {
+		osutil.HideConsole()
+	}
+
+	if c.HomeDir != "" {
+		if c.ConfDir != "" {
+			return fmt.Errorf("--home must not be used together with --config")
+		}
+		c.ConfDir = c.HomeDir
+	}
+	if c.ConfDir == "" {
+		c.ConfDir = locations.GetBaseDir(locations.ConfigBaseDir)
+	}
+
+	// Support reading the password from a pipe or similar
+	if c.GUIPassword == "-" {
+		reader := bufio.NewReader(os.Stdin)
+		password, _, err := reader.ReadLine()
+		if err != nil {
+			return fmt.Errorf("Failed reading GUI password: %w", err)
+		}
+		c.GUIPassword = string(password)
+	}
+
+	if err := Generate(c.ConfDir, c.GUIUser, c.GUIPassword, c.NoDefaultFolder); err != nil {
+		return fmt.Errorf("Failed to generate config and keys: %w", err)
+	}
+	return nil
+}
+
+func Generate(confDir, guiUser, guiPassword string, noDefaultFolder bool) error {
+	dir, err := fs.ExpandTilde(confDir)
+	if err != nil {
+		return err
+	}
+
+	if err := syncthing.EnsureDir(dir, 0700); err != nil {
+		return err
+	}
+	locations.SetBaseDir(locations.ConfigBaseDir, dir)
+
+	var myID protocol.DeviceID
+	certFile, keyFile := locations.Get(locations.CertFile), locations.Get(locations.KeyFile)
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err == nil {
+		log.Println("WARNING: Key exists; will not overwrite.")
+	} else {
+		cert, err = syncthing.GenerateCertificate(certFile, keyFile)
+		if err != nil {
+			return fmt.Errorf("create certificate: %w", err)
+		}
+	}
+	myID = protocol.NewDeviceID(cert.Certificate[0])
+	log.Println("Device ID:", myID)
+
+	cfgFile := locations.Get(locations.ConfigFile)
+	var cfg config.Wrapper
+	if _, err := os.Stat(cfgFile); err == nil {
+		if guiUser == "" && guiPassword == "" {
+			log.Println("WARNING: Config exists; will not overwrite.")
+			return nil
+		}
+
+		if cfg, _, err = config.Load(cfgFile, myID, events.NoopLogger); err != nil {
+			return fmt.Errorf("load config: %w", err)
+		}
+	} else {
+		if cfg, err = syncthing.DefaultConfig(cfgFile, myID, events.NoopLogger, noDefaultFolder); err != nil {
+			return fmt.Errorf("create config: %w", err)
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go cfg.Serve(ctx)
+	defer cancel()
+
+	guiCfg := cfg.GUI()
+	waiter, err := cfg.Modify(func(cfg *config.Configuration) {
+		if changed := updateGUIAuthentication(&guiCfg, guiUser, guiPassword); changed {
+			cfg.GUI = guiCfg
+		}
+	})
+	if err != nil {
+		return fmt.Errorf("modify config: %w", err)
+	}
+
+	waiter.Wait()
+	if err := cfg.Save(); err != nil {
+		return fmt.Errorf("save config: %w", err)
+	}
+	return nil
+}
+
+func updateGUIAuthentication(guiCfg *config.GUIConfiguration, guiUser, guiPassword string) bool {
+	changed := false
+	if guiUser != "" && guiCfg.User != guiUser {
+		guiCfg.User = guiUser
+		log.Println("Updated GUI authentication user name:", guiUser)
+		changed = true
+	}
+
+	if guiPassword != "" && guiCfg.Password != guiPassword {
```

Comment:
This is comparing a hash against a plain-text password, isn't it?  
Probably the field should be named `Password` but `PasswordHash`.

Reply:
You're right on both points.  I personally think that renaming is not worth the effort.

The comparison is a special case, because I wanted to skip writing if the password didn't actually change.  But the hash changes randomly (salted), so this is the only case we can actually detect.  If you think it's not useful, we can drop it though.

Reply:
What would you detect? The hash being equal to the plain-text password isn't something that should ever happen, and if it does then because someone manually added it to the config, in which case it should also be replaced.

Reply:
I see your concern and that makes sense to avoid.  My thought was about the opposite case, supplying an already hashed password.

Reply:
Ah right, now I get it - if for some arcane reason the user gets the hash from config and passes it to this function, it should just leave the password alone. Yeah I guess why not.

Reply:
> and if it does then because someone manually added it to the config, in which case it should also be replaced.

That case is really not nice though.  I'm undecided whether leaving a (wrongfully) unhashed password in the config when setting it again is wiser than trying to cater for this very hypothetical corner case...